### PR TITLE
Migrate to api v2

### LINKED ISSUE #127: nothing will play last few days
I've been trying with both standard and alt sources but I'm forever getting no steam available on anything I try to play.  Is it just me?

### LINKED ISSUE #128: Nothing will play except Black Clover
The only one that plays is Black Clover!  Everybody else just clicking icon will give errors. SEARCH ALT will run after it pressed a couple times. And I seen only one search alt give me a done!

### LINKED ISSUE #131: Unable to browse latest
Selecting Latest gives an error message, also notable many of the recent releases are unable to find a source.

### LINKED ISSUE #132: Is this dead? I can't do anything.
Can't get logs due to not knowing where the heck they are through ADBLink.

### LINKED ISSUE #133: Can't use the addon
Using the lastest version of kodi (18.6 I think).
Happens for like 2 weeks now,no matter the show.
I can enter the addon and use the search,but that's about it.
Here's the log:

> 
2020-03-19 11:31:41.840 T:11112   ERROR: XFILE::CDirectory::GetDirectory - Error getting plugin://plugin.video.wonderfulsubs/search/code geass/1
2020-03-19 11:31:41.852 T:6528   ERROR: CGUIMediaWindow::GetDirectory(plugin://plugin.video.wonderfulsubs/search/code geass/1) failed
2020-03-19 11:31:43.237 T:6528 WARNING: Repository add-on repository.ptom uses plain HTTP for add-on downloads in path http://raw.githubusercontent.com/ptom98/PTOM/master/ - this is insecure and will make your Kodi installation vulnerable to attacks if enabled!
2020-03-19 11:34:32.851 T:6528 WARNING: Previous line repeats 2 times.
2020-03-19 11:34:32.851 T:6528   ERROR: Control 50 in window 10025 has been asked to focus, but it can't
2020-03-19 11:34:35.451 T:14792   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.ValueError'>
                                            Error Contents: No JSON object could be decoded
                                            Traceback (most recent call last):
                                              File "D:\Program Files\Kodi\portable_data\addons\plugin.video.wonderfulsubs\default.py", line 292, in <module>
                                                router_process(control.get_plugin_url(), control.get_plugin_params())
                                              File "D:\Program Files\Kodi\portable_data\addons\plugin.video.wonderfulsubs\resources\lib\ui\router.py", line 68, in router_process
                                                return route_obj.func(payload, params)
                                              File "D:\Program Files\Kodi\portable_data\addons\plugin.video.wonderfulsubs\default.py", line 230, in SEARCH_PAGES
                                                return control.draw_items(_BROWSER.search_site(query, int(page)), draw_cm=draw_cm)
                                              File "D:\Program Files\Kodi\portable_data\addons\plugin.video.wonderfulsubs\resources\lib\WonderfulSubsBrowser.py", line 289, in search_site
                                                return self._process_anime_view(url, data, "search/%s/%%d" % search_string, page)
                                              File "D:\Program Files\Kodi\portable_data\addons\plugin.video.wonderfulsubs\resources\lib\WonderfulSubsBrowser.py", line 107, in _process_anime_view
                                                json_resp = self._json_request(url, data)
                                              File "D:\Program Files\Kodi\portable_data\addons\plugin.video.wonderfulsubs\resources\lib\WonderfulSubsBrowser.py", line 98, in _json_request
                                                response = json.loads(self._get_request(url, data))
                                              File "D:\Program Files\Kodi\system\python\Lib\json\__init__.py", line 339, in loads
                                                return _default_decoder.decode(s)
                                              File "D:\Program Files\Kodi\system\python\Lib\json\decoder.py", line 364, in decode
                                                obj, end = self.raw_decode(s, idx=_w(s, 0).end())
                                              File "D:\Program Files\Kodi\system\python\Lib\json\decoder.py", line 382, in raw_decode
                                                raise ValueError("No JSON object could be decoded")
                                            ValueError: No JSON object could be decoded
                                            -->End of Python script error report<--
2020-03-19 11:34:35.480 T:10008   ERROR: XFILE::CDirectory::GetDirectory - Error getting plugin://plugin.video.wonderfulsubs/search/code geass/1
2020-03-19 11:34:35.485 T:6528   ERROR: CGUIMediaWindow::GetDirectory(plugin://plugin.video.wonderfulsubs/search/code geass/1) failed

### PR DESCRIPTION
**Pull request by nalch**
1. Move base url and api base into own module
2. Use new api base where applicable
3. Add login settings
4. Add login/logout functions
5. Add automatic (re)login after forbidden responses
6. Minor pycodestyle fixes

fixes #127, fixes #128, fixes #131, fixes #132, fixes #133

### CODE REVIEW

**Comment by doko-desuka** [comment]
Wow, this looks like a lot of work! Thanks for all this @nalch. 

I think this official WS add-on could use a cleanup in some places.  
The [/resources/lib/ui/embed_extractor.py](https://github.com/DxCx/plugin.video.wonderfulsubs/blob/master/resources/lib/ui/embed_extractor.py) script for example, it has some extractors for sites that don't exist anymore like RapidVideo. Another point is the fact that it requests series data multiple times when it could cache it to memory (see usages of the [_get_anime_info](https://github.com/DxCx/plugin.video.wonderfulsubs/blob/master/resources/lib/WonderfulSubsBrowser.py#L203) function).  
I mean, if you already have the data, put it in a Kodi memory property and keep it there ([like it's done in here](https://github.com/doko-desuka/plugin.video.wsbeta/blob/9ac74bdf3f898f96ad79da6ec5076fcf89a9b658/Lib/Plugin.py#L714-L736)).

**Comment by doko-desuka** [comment]
For everyone coming in, **you can download this zip here to install this updated WonderfulSubs version by Nalch**: [plugin.video.wonderfulsubs.zip](https://github.com/DxCx/plugin.video.wonderfulsubs/files/4364781/plugin.video.wonderfulsubs.zip) *  
If any more commits are made after this comment, this zip will become obsolete and should be ignored.

\* That zip is just [the archive of his fork](https://github.com/nalch/plugin.video.wonderfulsubs/archive/migrate-to-api-v2.zip) but with the folder inside it renamed to `plugin.video.wonderfulsubs`.

**Comment by Ichijoe2112** [comment]
Just tried this, and it's not working. This said. It would seem like the entire WonderfulSubs site is um dead. Going there points you to use the 'beta.WonderfulSubs.com' location. That on first inspection looks more promising. But, nothing seems to be quite baked there as well. Hopefully these issues will be resolved soon. But, I got a feeling things are staying down again for a while T_T

**Comment by nalch** [comment]
> Just tried this, and it's not working. This said. It would seem like the entire WonderfulSubs site is um dead. Going there points you to use the 'beta.WonderfulSubs.com' location. That on first inspection looks more promising. But, nothing seems to be quite baked there as well. Hopefully these issues will be resolved soon. But, I got a feeling things are staying down again for a while T_T
> 
> But, I thank the Developers use of their time. And, having no intentions of disrespecting it either. Even if nothing much has changed. But, as stated, I believe these troubles are all on the other side of the ether right now.

That is weird. Just watched a bit with the new version.
The pull request version uses the beta version of the site (or more precisely api version v2), so that shouldn't be a problem. If you can upload the kodi log, I'll have a look, what could have gone wrong.

**Comment by DxCx** [comment]
Awsome work @nalch thanks alot for your hard work!
Ill review it ASAP :)

**Comment by NerdNational** [comment]
is there away of getting the plugin to work on the firestick?

**Comment by Ichijoe2112** [comment]
I haven't used a Fire Device in some time. So I might be out of touch. But, assuming things haven't changed much. You should be a side / down load(er), from the Amazon Appstore. That you will endup pluging the URL to Kodi into. Which will allow you to download, and install it. Its a shame this still has to be said in 2020. But, Google really is your best Friend in such matters, use him.

**Comment by NerdNational** [comment]
That isn’t the issue I am having I am having issues with the videos not working as log shows

**Comment by doko-desuka** [comment]
@NerdNational are you able to watch it on the website? because some old / obscure shows simply don't have any working sources anymore.  
They still reference Mangostream / RapidVideo, video hosters that don't exist anymore.

If what you're trying to watch doesn't have the CR server then I wouldn't put much hope into it.

**Comment by doko-desuka** [comment]
* Streamango, not Mangostream lol

**Comment by NerdNational** [comment]
Yup I can watch it on the site with no issues I cannot even load the category

**Comment by Ichijoe2112** [comment]
> > Just tried this, and it's not working. This said. It would seem like the entire WonderfulSubs site is um dead. Going there points you to use the 'beta.WonderfulSubs.com' location. That on first inspection looks more promising. But, nothing seems to be quite baked there as well. Hopefully these issues will be resolved soon. But, I got a feeling things are staying down again for a while T_T
> > But, I thank the Developers use of their time. And, having no intentions of disrespecting it either. Even if nothing much has changed. But, as stated, I believe these troubles are all on the other side of the ether right now.
> 
> That is weird. Just watched a bit with the new version.
> The pull request version uses the beta version of the site (or more precisely api version v2), so that shouldn't be a problem. If you can upload the kodi log, I'll have a look, what could have gone wrong.

Ok I like to think I'm on top of things. But, I'm willing to admit stupidity on my part. The Zip Files as given... Do I need to treat these is any other special way, other then the usual method of 'installing as zip file' (i.e. System -> Settings -> Add-ons -> Install from zip file)?
Because this is how I did it, on my Shield TV running Kodi v18.6.2

I was also trying to access the Site from my Tablet, as I was requested to log back into my old account. Which was Not workong for me. Henne the Beta Website which did allow me to create a working account. Which this Add-on seems to also accept. No more nagging about needing to log in. But,  (and, it may be a tablet thing), but, nothing was apparently working on the actual website over Crome either.

**Comment by NerdNational** [comment]
???

**Comment by NerdNational** [comment]
Same for me

**Comment by NerdNational** [comment]
I have no idea what that means tbh XD

**Comment by Ichijoe2112** [comment]
Ok Tried somethings differently, and have since found some success. My main point of failure is in using 'Latest'.  As per my want, if I instead look into the 'Airing' channel. Things begin to work. But, I'm to lasy to read subs, and want to watch the dubbed stuff. And, that may well be all still hit, and miss at this point. As I was trying to start a dubbed ep of new LotGH on my Samsung Tablet last night, and the Site was clearly not having any of it.

**Comment by doko-desuka** [comment]
Can you test it with this add-on? https://github.com/doko-desuka/plugin.video.wsbeta

File: resources/lib/WonderfulSubsBrowser.py [hunk 1]
```diff
@@ -97,14 +99,14 @@ def _handle_watchlist_paging(self, results, base_url, page):
...
-    def _json_request(self, url, data):
-        response = json.loads(self._get_request(url, data))
-        if response["status"] != 200:
-            raise Exception("Request %s returned with error code %d" % (url,
-                                                                        response["status"]))
+    def _json_request(self, url, data, reauth=True):
+        response = self._get_request(url, data)
 
-        return response["json"]
+        if self._response_forbidden(response) and reauth:
```
Comment (DxCx) [inline resources/lib/WonderfulSubsBrowser.py]: Yeah this response foribidden is part of this class, not the base..

File: resources/lib/WonderfulSubsBrowser.py [hunk 2]
```diff
@@ -97,14 +99,14 @@ def _handle_watchlist_paging(self, results, base_url, page):
...
+            login()
+            return self._json_request(url, data, reauth=False)
```
Comment (DxCx) [inline resources/lib/WonderfulSubsBrowser.py]: Dont you need to just tell xbmc to referesh?
Also, i think there was another place which tried to handle relogin..

File: resources/lib/WonderfulSubsBrowser.py (hunk 1 above)
Comment (DxCx) [reply resources/lib/WonderfulSubsBrowser.py]: And you can still get non-200 which is not foribbdden

File: resources/lib/__init__.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+from .login import login, logout
+from .ui.router import route
+
+
+@route('wonderful_login')
```
Comment (DxCx) [inline resources/lib/__init__.py]: Why those are here and not under default?

File: resources/settings.xml [hunk 4]
```diff
@@ -5,6 +5,13 @@
 		<setting id="reverseorder" label="40002" type="labelenum" values="Ascending|Descending" default="Descending"/>
 		<setting id="searchhistory" label="40003" type="labelenum" values="Yes|No" default="Yes"/>
 
+		<setting type="lsep" label="Login"/>
```
Comment (DxCx) [inline resources/settings.xml]: So looks like we need to remove wonderful subs from the playlist flavor (as everone is logged in by default) and handle different flavor only if set..

Comment (DxCx) [reply resources/settings.xml]: Also note that some changes leaked from this commit into previous one, like strings and forbiden/use token

File: resources/lib/AniListBrowser.py [hunk 5]
```diff
@@ -1,11 +1,11 @@
 import itertools
```
Comment (DxCx) [inline resources/lib/AniListBrowser.py]: `Minor pycodestyle fixes` - while i am in favor of using tooling i think that commit alone deserves its own PR.
Also, if using such tool, the configuration file should be shared in the repo (to make sure everyone is aligned), so please add such file to git (tox.ini?)

File: resources/lib/WonderfulSubsBrowser.py [hunk 6]
```diff
@@ -301,39 +308,39 @@ def get_by_letter(self, letter, page = 1):
...
+        url = self._to_url("{}/media/latest".format(API_BASE))
         return self._process_anime_view(url, data, "latest/%d", page)
 
     def get_random(self, page=1):
         data = {
             "count": self._RESULTS_PER_SEARCH_PAGE,
             "index": (page-1) * self._RESULTS_PER_SEARCH_PAGE,
         }
-        url = self._to_url("api/media/random")
+        url = self._to_url("{}/media/random".format(API_BASE))
```
[the remark this answers is not in the document]
Comment (nalch) [reply resources/lib/WonderfulSubsBrowser.py]: I didn't want to change the base class behaviour, because I didn't know, where it was used. But I'll just set ```_BASE_URL``` to BASE_URL + API_BASE and we should be set :)

File: resources/lib/ui/http.py [hunk 7]
```diff
@@ -155,6 +157,10 @@ def __send_request(session, url, data=None, set_request=None, head=False):
         'allow_redirects': True,
     }
 
+    token = control.getSetting('login.token')
```
[the remark this answers is not in the document]
Comment (nalch) [reply resources/lib/ui/http.py]: Not really sure, where to put it. Do you mean to pass it from WonderfulSubsBrowser into the function?

File: resources/lib/ui/embed_extractor.py [hunk 8]
```diff
@@ -8,6 +8,8 @@
 import time
 from bs4 import BeautifulSoup
 
+from ..constants import API_BASE, BASE_URL
```
[the remark this answers is not in the document]
Comment (nalch) [reply resources/lib/ui/embed_extractor.py]: Puh. Not even constants? Should I copy the urls in here?

File: resources/lib/ui/BrowserBase.py [hunk 9]
```diff
@@ -21,3 +21,11 @@ def _get_request(self, url, data=None, set_request=None):
         if data:
             url = "%s?%s" % (url, urllib.urlencode(data))
         return self._send_request(url, None, set_request)
+
+    def _response_forbidden(self, response):
```
[the remark this answers is not in the document]
Comment (nalch) [reply resources/lib/ui/BrowserBase.py]: That's usable by all subclasses most notably WonderfulSubsBrowser. It's basically used to see, that we should (re)login with the stored credentials.

At the place I want to use it (_json_request), we only get the text of the http response :/ (coming from _send_request), so no status code unfortunately.

File: resources/lib/WonderfulSubsBrowser.py (hunk 1 above)
Comment (nalch) [reply resources/lib/WonderfulSubsBrowser.py]: This part only deals with actual forbidden responses. Everything else should still be an error. It just makes it easy to (re)login, if credentials are stored in the settings, but no token exists or it needs refreshing.

File: resources/lib/WonderfulSubsBrowser.py (hunk 2 above)
Comment (nalch) [reply resources/lib/WonderfulSubsBrowser.py]: My knowledge about xmbc is not sufficient for that :)

File: resources/lib/__init__.py (hunk 3 above)
Comment (nalch) [reply resources/lib/__init__.py]: Good point. Didn't see default.py, so I'll move it there!

File: resources/settings.xml (hunk 4 above)
Comment (nalch) [reply resources/settings.xml]: Seems like you can browse some lists without being logged in, so not sure, if we want to remove the watchlist flavor. Probably a good idea.

File: resources/lib/AniListBrowser.py (hunk 5 above)
Comment (nalch) [reply resources/lib/AniListBrowser.py]: Yeah, was conflicted myself, but those are just a subset of the pycodestyle rules my IDE showed :/ The inner monk wanted the empty lines.

Comment (DxCx) [reply resources/lib/AniListBrowser.py]: Im perfectly fine with your inner monk, it just make sense, i just want the config to be shared so other monks woth get nuts ;)

File: resources/lib/ui/BrowserBase.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (DxCx) [reply resources/lib/ui/BrowserBase.py]: Yes but as mentioned, forbiden response is part of ws server, its not generic for base browser

File: resources/lib/ui/http.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (DxCx) [reply resources/lib/ui/http.py]: Yep, token is generic to http, control isnt

File: resources/lib/ui/embed_extractor.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (DxCx) [reply resources/lib/ui/embed_extractor.py]: Yeah i prefere this to be compeltly seperated as this directory is shared between plugins

File: resources/lib/WonderfulSubsBrowser.py (hunk 1 above)
Comment (DxCx) [reply resources/lib/WonderfulSubsBrowser.py]: Yes but it should be with on the same class

File: resources/lib/WonderfulSubsBrowser.py (hunk 2 above)
Comment (DxCx) [reply resources/lib/WonderfulSubsBrowser.py]: Ill need to referesh my memory but i think there is a function to halt render and just referesh the same addon url

File: resources/settings.xml (hunk 4 above)
Comment (DxCx) [reply resources/settings.xml]: You need it to play? What gives browse without play?

File: resources/lib/WonderfulSubsBrowser.py (hunk 2 above)
Comment (sammydc77) [reply resources/lib/WonderfulSubsBrowser.py]: II was exploring today, on another issue, and came across: Kodi Documentation website [ https://codedocs.xyz/AlwinEsch/kodi/index.html ].  It appears to be for v19.0, so I do not know if you can find something useful here.  nalch, I thought this might help.  

Also I
 tried your version yesterday and watched 22/7 sub [from Currently Airing].  I did not get any subs, on screen.  I reviewed the log, and found WS could not find sub.  I apologize, but I did not copy the segment of the log referencing this issue...  Sorry, just thought I would mention this.

I really appreciate all you are doing!

File: resources/lib/WonderfulSubsBrowser.py (hunk 1 above)
Comment (nalch) [reply resources/lib/WonderfulSubsBrowser.py]: Yeah, now I got it. Makes perfect sense! Thanks!

File: resources/lib/ui/embed_extractor.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nalch) [reply resources/lib/ui/embed_extractor.py]: Ah, good information. That makes sense.

File: resources/settings.xml (hunk 4 above)
Comment (nalch) [reply resources/settings.xml]: Fair point. I usually just log in when I have to, but I'll try to put it on load.

File: resources/lib/__init__.py [hunk 10]
```diff
@@ -0,0 +1,4 @@
+
```
[the remark this answers is not in the document]
Comment (nalch) [reply resources/lib/__init__.py]: Done :)

**Comment by nalch** [comment]
> Also, if you please can rebase and squash properly so i can use the commits as is and not squash all togather

Not sure, what you mean with that. I usually have the commits separate for easier reviewing and then squash them when merging. What do you prefer to do?

**Comment by DxCx** [comment]
Everything is reviewed and can be prepared for a merge ;)

**Comment by Serg86** [comment]
Is the upcoming new version still compatible with Kodi 17.6? I am using RecalBox 6.1.1 and unfortunately, their Kodi version is always way older than the current Kodi release.

I much prefer RecalBox to RetroPie and even though Kodi introduced RetroPlayer in version 18, it's far from great, the input lag alone is horrendous. Maybe Kodi 19 or 20 will be better but so far, RecalBox it is, which unfortunately means Kodi 17.6 for the foreseeable future.

**Comment by DxCx** [comment]
@nalch should i expect another drop soon? Or do you want me to take it from this point to merge? I think that espeically now (covid-19) people would need this plugin even more

**Comment by nalch** [comment]
> @nalch should i expect another drop soon? Or do you want me to take it from this point to merge? I think that espeically now (covid-19) people would need this plugin even more

Hey, I'm probably not able to add something until late next week so feel free to chime in :)

**Comment by DxCx** [comment]
please open a PR with pycodestyle fixes (including its config)
